File: verification/scripts/python/feature_importance/test_feature_importance.py
# ============================================================================
# Feature Importance Test Reference Implementation (Python with statsmodels)
# ============================================================================
# This script generates reference values for feature importance using Python's
# statsmodels and scikit-learn libraries. The test validates that the Rust
# implementation matches Python's behavior for feature importance metrics.
#
# Required packages:
#   pip install numpy pandas scipy statsmodels scikit-learn
#
# Usage:
#   python test_feature_importance.py --csv [csv_path] --output-dir [output_dir]
#   Args:
#     --csv        - Path to CSV file (first col = response, rest = predictors)
#                    Default: ../../datasets/csv/mtcars.csv
#     --output-dir - Path to output directory
#                    Default: ../../results/python
# ============================================================================

import argparse
import numpy as np
import pandas as pd
import json
import sys
from pathlib import Path
from typing import List, Dict, Any

# Native Python libraries for reference implementation
import statsmodels.api as sm
from statsmodels.regression.linear_model import OLS
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.metrics import r2_score
from scipy import stats


class NumpyEncoder(json.JSONEncoder):
    """JSON encoder for numpy types."""
    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        if isinstance(obj, np.floating):
            return float(obj)
        if isinstance(obj, np.ndarray):
            return obj.tolist()
        if isinstance(obj, np.float64):
            return float(obj)
        if isinstance(obj, np.int64):
            return int(obj)
        return super(NumpyEncoder, self).default(obj)


def convert_categorical_to_numeric(data, dataset_name):
    """Convert categorical columns to numeric representations (0-based factor codes)."""
    non_numeric_cols = data.select_dtypes(exclude=[np.number]).columns.tolist()

    if len(non_numeric_cols) > 0:
        print(f"INFO: Dataset '{dataset_name}' contains non-numeric columns: "
              f"{', '.join(non_numeric_cols)}")
        print("Converting categorical variables to numeric factor codes...")

        for col in non_numeric_cols:
            # First try to convert to numeric directly (handles numeric strings)
            numeric_vals = pd.to_numeric(data[col], errors='coerce')
            if not numeric_vals.isna().any():
                # All values converted successfully - use numeric values
                data[col] = numeric_vals
            else:
                # Contains non-numeric strings - convert to factor codes (0-based)
                codes, _ = pd.factorize(data[col])
                data[col] = codes.astype(float)
                print(f"  Column '{col}' encoded as 0-based factor codes")

    return data


# =============================================================================
# Standardized Coefficients (using statsmodels)
# =============================================================================

def validate_standardized_coefficients(
    df: pd.DataFrame,
    predictors: List[str],
    response: str
) -> Dict[str, Any]:
    """
    Compute standardized coefficients using statsmodels.
    """
    X = df[predictors].values
    y = df[response].values

    # Fit OLS model
    X_with_const = sm.add_constant(X)
    model = OLS(y, X_with_const).fit()

    # Get coefficients (excluding intercept)
    coefficients = model.params[1:]  # Skip intercept

    # Compute standard deviations (using scipy.stats for compatibility)
    x_stds = stats.tstd(X, ddof=X.shape[0] - 1, axis=0)
    y_std = stats.tstd(y, ddof=len(y) - 1)

    # Standardized coefficients: beta* = beta * (sigma_x / sigma_y)
    std_coefs = coefficients * (x_stds / y_std)

    result = {
        "variable_names": predictors,
        "standardized_coefficients": std_coefs.tolist(),
        "y_std": float(y_std),
        "raw_coefficients": coefficients.tolist()
    }

    return result


# =============================================================================
# VIF Ranking (using statsmodels)
# =============================================================================

def validate_vif_ranking(
    df: pd.DataFrame,
    predictors: List[str],
    response: str
) -> Dict[str, Any]:
    """
    Compute VIF values using statsmodels.
    """
    X = df[predictors].values
    X_with_const = sm.add_constant(X)

    vif_data = []
    for i, var_name in enumerate(predictors):
        # VIF calculation using statsmodels
        vif_val = variance_inflation_factor(X_with_const, i + 1)

        # Handle edge cases
        if np.isinf(vif_val):
            rsquared = 1.0
        elif vif_val > 0:
            rsquared = max(0.0, 1.0 - 1.0 / vif_val)
        else:
            rsquared = 0.0

        # Determine interpretation
        if vif_val < 5:
            interpretation = "Low multicollinearity"
        elif vif_val < 10:
            interpretation = "Moderate multicollinearity"
        else:
            interpretation = "High multicollinearity"

        vif_data.append({
            "variable": var_name,
            "vif": float(vif_val) if np.isfinite(vif_val) else float('inf'),
            "rsquared": rsquared,
            "interpretation": interpretation
        })

    # Sort by VIF (ascending = least redundant first)
    vif_sorted = sorted(vif_data, key=lambda x: x['vif'] if x['vif'] != float('inf') else float('inf'))

    return {
        "variable_names": [v['variable'] for v in vif_sorted],
        "vif_values": [v['vif'] for v in vif_sorted],
        "ranking": vif_sorted
    }


# =============================================================================
# Linear SHAP Values (exact, closed-form)
# =============================================================================

def validate_linear_shap(
    df: pd.DataFrame,
    predictors: List[str],
    response: str
) -> Dict[str, Any]:
    """
    Compute exact SHAP values for linear regression using statsmodels.
    """
    X = df[predictors].values
    y = df[response].values

    # Fit OLS model
    X_with_const = sm.add_constant(X)
    model = OLS(y, X_with_const).fit()

    # Get coefficients
    coefficients = model.params[1:]  # Skip intercept
    intercept = model.params[0]

    # Compute means of predictors
    X_means = X.mean(axis=0)

    n_obs = len(y)
    n_features = len(predictors)

    # Compute SHAP values matrix
    shap_values = []
    for i in range(n_obs):
        obs_shap = []
        for j in range(n_features):
            shap_val = coefficients[j] * (X[i, j] - X_means[j])
            obs_shap.append(shap_val)
        shap_values.append(obs_shap)

    # Compute mean absolute SHAP per feature
    shap_array = np.array(shap_values)
    mean_abs_shap = np.abs(shap_array).mean(axis=0).tolist()

    return {
        "variable_names": predictors,
        "shap_values": shap_values,
        "base_value": float(intercept),
        "mean_abs_shap": mean_abs_shap
    }


# =============================================================================
# Permutation Importance (using scikit-learn)
# =============================================================================

def validate_permutation_importance(
    df: pd.DataFrame,
    predictors: List[str],
    response: str,
    n_permutations: int = 10,
    random_seed: int = 42
) -> Dict[str, Any]:
    """
    Compute permutation importance using scikit-learn.
    """
    np.random.seed(random_seed)

    X = df[predictors].values
    y = df[response].values

    # Fit baseline model
    from sklearn.linear_model import LinearRegression
    from sklearn.base import clone

    model = LinearRegression()
    model.fit(X, y)
    baseline_r2 = r2_score(y, model.predict(X))

    n_obs, n_features = X.shape
    importance_scores = []

    for j in range(n_features):
        perm_r2_sum = 0.0

        for _ in range(n_permutations):
            X_permuted = X.copy()

            # Shuffle column j
            perm_col = X_permuted[:, j].copy()
            np.random.shuffle(perm_col)
            X_permuted[:, j] = perm_col

            # Fit on permuted data and compute R²
            perm_model = clone(model)
            perm_model.fit(X_permuted, y)
            perm_r2 = r2_score(y, perm_model.predict(X_permuted))

            perm_r2_sum += baseline_r2 - perm_r2

        importance_scores.append(perm_r2_sum / n_permutations)

    return {
        "variable_names": predictors,
        "importance": importance_scores,
        "baseline_score": float(baseline_r2),
        "n_permutations": n_permutations,
        "seed": random_seed
    }


def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(
        description="Generate feature importance reference values"
    )
    parser.add_argument(
        "--csv",
        default="../../datasets/csv/mtcars.csv",
        help="Path to CSV file (first col = response, rest = predictors)"
    )
    parser.add_argument(
        "--output-dir",
        default="../../results/python",
        help="Path to output directory"
    )

    args = parser.parse_args()

    # Resolve paths
    csv_path = Path(args.csv) if Path(args.csv).is_absolute() else Path.cwd() / args.csv
    output_dir = Path(args.output_dir) if Path(args.output_dir).is_absolute() else Path.cwd() / args.output_dir

    # Validate CSV path
    if not csv_path.exists():
        print(f"ERROR: CSV file not found: {csv_path}")
        sys.exit(1)

    # Extract dataset name from filename
    dataset_name = csv_path.stem

    print(f"Running feature importance test on dataset: {dataset_name}")

    # Load data
    data = pd.read_csv(csv_path)
    data = convert_categorical_to_numeric(data, dataset_name)

    # Use first column as response, rest as predictors
    response = data.columns[0]
    predictors = [col for col in data.columns if col != response]

    print(f"Predictors: {predictors}")
    print(f"Response: {response}")
    print(f"Observations: {len(data)}")

    # Skip if too few predictors
    if len(predictors) < 1:
        print(f"Skipping {dataset_name}: insufficient predictors")
        sys.exit(1)

    n_features = len(predictors)

    # Single predictor case - skip VIF (requires at least 2 predictors)
    has_vif = n_features >= 2

    result = {
        "test": "feature_importance",
        "method": "statsmodels",
        "dataset": dataset_name,
        "n": len(data),
        "k": n_features,
        "variable_names": predictors,
        "response": response
    }

    # 1. Standardized Coefficients
    print("\n[1/4] Standardized Coefficients...")
    try:
        std_coefs = validate_standardized_coefficients(data, predictors, response)
        result["standardized_coefficients"] = std_coefs
        print(f"   Variables: {std_coefs['variable_names']}")
        print(f"   Standardized Coefs: {[f'{x:.4f}' for x in std_coefs['standardized_coefficients']]}")
    except Exception as e:
        print(f"   Error: {e}")
        import traceback
        traceback.print_exc()

    # 2. VIF Ranking (skip for single predictor)
    if has_vif:
        print("\n[2/4] VIF Ranking...")
        try:
            vif_result = validate_vif_ranking(data, predictors, response)
            result["vif_ranking"] = vif_result
            for v in vif_result['ranking'][:5]:
                print(f"   {v['variable']}: VIF={v['vif']:.2f} ({v['interpretation']})")
        except Exception as e:
            print(f"   Error: {e}")
            import traceback
            traceback.print_exc()
    else:
        print("\n[2/4] VIF Ranking... Skipped (only 1 predictor)")

    # 3. Linear SHAP
    print("\n[3/4] Linear SHAP...")
    try:
        shap_result = validate_linear_shap(data, predictors, response)
        result["shap"] = shap_result
        print(f"   Mean |SHAP|: {[f'{x:.4f}' for x in shap_result['mean_abs_shap']]}")
    except Exception as e:
        print(f"   Error: {e}")
        import traceback
        traceback.print_exc()

    # 4. Permutation Importance
    print("\n[4/4] Permutation Importance...")
    try:
        perm_result = validate_permutation_importance(data, predictors, response, n_permutations=10)
        result["permutation_importance"] = perm_result
        print(f"   Importance scores: {[f'{x:.4f}' for x in perm_result['importance']]}")
    except Exception as e:
        print(f"   Error: {e}")
        import traceback
        traceback.print_exc()

    # Ensure output directory exists
    output_dir.mkdir(parents=True, exist_ok=True)

    # Write output
    output_file = output_dir / f"{dataset_name}_feature_importance.json"
    with open(output_file, 'w') as f:
        json.dump(result, f, cls=NumpyEncoder, indent=2)

    print(f"\nWrote: {output_file.absolute()}")


if __name__ == "__main__":
    main()

File: verification/scripts/r/feature_importance/test_feature_importance.R
# ============================================================================
# Feature Importance Test Reference Implementation (R)
# ============================================================================
# This script generates reference values for feature importance using R's
# native libraries (stats, car). The test validates that the Rust
# implementation matches R's behavior for feature importance metrics.
#
# Required packages:
#   install.packages(c("stats", "car", "jsonlite"))
#
# Usage:
#   Rscript test_feature_importance.R [csv_path] [output_dir]
#   Args:
#     csv_path    - Path to CSV file (first col = response, rest = predictors)
#                   Default: ../../../datasets/csv/mtcars.csv
#     output_dir  - Path to output directory
#                   Default: ../../../results/r
# ============================================================================

# Get script directory and resolve relative paths
args_all <- commandArgs(trailingOnly = FALSE)
script_arg <- args_all[grep("^--file=", args_all)]
if (length(script_arg) > 0) {
  script_path <- dirname(sub("--file=", "", script_arg))
  original_wd <- getwd()
  setwd(script_path)
} else {
  original_wd <- getwd()
}

# Add user library to path
user_lib <- file.path(Sys.getenv("USERPROFILE"), "Documents", "R", "win-library", "4.4")
if (dir.exists(user_lib)) {
  .libPaths(c(user_lib, .libPaths()))
}

# Load required packages
suppressPackageStartupMessages({
  library(stats)
  library(car)
  library(jsonlite)
})

# =============================================================================
# Data Helper Functions
# =============================================================================

convert_categorical_to_numeric <- function(data, dataset_name) {
  non_numeric_cols <- names(data)[sapply(data, function(x) !is.numeric(x))]

  if (length(non_numeric_cols) > 0) {
    cat(paste0("INFO: Dataset '", dataset_name, "' contains non-numeric columns: ",
               paste(non_numeric_cols, collapse = ", "), "\n"))
    cat("Converting categorical variables to numeric representations...\n")

    for (col in non_numeric_cols) {
      if (is.factor(data[[col]])) {
        data[[col]] <- as.numeric(data[[col]])
      } else if (is.character(data[[col]])) {
        temp_numeric <- as.numeric(data[[col]])
        if (any(is.na(temp_numeric))) {
          data[[col]] <- as.numeric(as.factor(data[[col]]))
        } else {
          data[[col]] <- temp_numeric
        }
      }
    }
  }

  return(data)
}

# Resolve paths relative to original working directory
resolve_path <- function(path) {
  is_absolute <- grepl("^(/|[A-Za-z]:)", path)
  if (!is_absolute) {
    file.path(original_wd, path)
  } else {
    path
  }
}

# =============================================================================
# Standardized Coefficients (using base R)
# =============================================================================

validate_standardized_coefficients <- function(df, predictors, response) {
  formula_str <- paste(response, "~", paste(predictors, collapse = " + "))
  formula <- as.formula(formula_str)

  model <- lm(formula, data = df)

  # Get coefficients (excluding intercept)
  coefficients <- coef(model)[-1]
  variable_names <- names(coefficients)

  # Compute standard deviations using base R
  # Use drop=FALSE to ensure we always get a data frame (even for single column)
  x_stds <- sapply(df[, predictors, drop = FALSE], sd, na.rm = TRUE)
  y_std <- sd(df[[response]], na.rm = TRUE)

  # Standardized coefficients
  std_coefs <- coefficients * (x_stds / y_std)

  list(
    variable_names = variable_names,
    standardized_coefficients = as.numeric(std_coefs),
    y_std = y_std,
    raw_coefficients = as.numeric(coefficients)
  )
}

# =============================================================================
# VIF Ranking (using car package)
# =============================================================================

validate_vif_ranking <- function(df, predictors, response) {
  formula_str <- paste(response, "~", paste(predictors, collapse = " + "))
  formula <- as.formula(formula_str)

  model <- lm(formula, data = df)

  # VIF using car::vif
  vif_values <- car::vif(model)

  # Add interpretation
  interpretations <- sapply(vif_values, function(v) {
    if (v < 5) "Low multicollinearity"
    else if (v < 10) "Moderate multicollinearity"
    else "High multicollinearity"
  })

  vif_data <- data.frame(
    variable = names(vif_values),
    vif = as.numeric(vif_values),
    rsquared = 1 - 1/vif_values,
    interpretation = interpretations,
    stringsAsFactors = FALSE
  )

  # Sort by VIF ascending (least redundant first)
  vif_sorted <- vif_data[order(vif_data$vif), ]

  list(
    variable_names = vif_sorted$variable,
    vif_values = vif_sorted$vif,
    ranking = vif_sorted
  )
}

# =============================================================================
# Linear SHAP Values (exact, closed-form)
# =============================================================================

validate_linear_shap <- function(df, predictors, response) {
  formula_str <- paste(response, "~", paste(predictors, collapse = " + "))
  formula <- as.formula(formula_str)

  model <- lm(formula, data = df)

  # Get coefficients
  coefficients <- coef(model)[-1]
  intercept <- coef(model)[1]

  # Compute means of predictors
  X <- as.matrix(df[, predictors])
  X_means <- colMeans(X)

  # Compute SHAP values: SHAP_i = coef_i * (x_i - mean(x_i))
  # Handle single-predictor case
  if (length(predictors) == 1) {
    shap_values <- matrix(coefficients * (X - X_means), nrow = 1)
  } else {
    shap_values <- apply(X, 1, function(row) {
      coefficients * (row - X_means)
    })
  }

  # shap_values is a k × n matrix (k features × n observations)
  # Mean absolute SHAP per feature: mean across observations (rows)
  mean_abs_shap <- rowMeans(abs(shap_values))

  list(
    variable_names = predictors,
    shap_values = t(shap_values),
    base_value = as.numeric(intercept),
    mean_abs_shap = as.numeric(mean_abs_shap)
  )
}

# =============================================================================
# Permutation Importance (using base R)
# =============================================================================

validate_permutation_importance <- function(df, predictors, response,
                                            n_permutations = 10,
                                            random_seed = 42) {
  set.seed(random_seed)

  X <- as.matrix(df[, predictors])
  y <- df[[response]]

  # Fit baseline model
  model <- lm(y ~ X)
  baseline_r2 <- summary(model)$r.squared

  n_obs <- nrow(X)
  n_features <- ncol(X)
  importance_scores <- numeric(n_features)

  for (j in 1:n_features) {
    perm_r2_sum <- 0

    for (iter in 1:n_permutations) {
      X_permuted <- X
      # Shuffle column j
      perm_col <- X_permuted[, j]
      X_permuted[, j] <- sample(perm_col)

      # Fit on permuted data
      perm_model <- lm(y ~ X_permuted)
      perm_r2 <- summary(perm_model)$r.squared

      perm_r2_sum <- perm_r2_sum + (baseline_r2 - perm_r2)
    }

    importance_scores[j] <- perm_r2_sum / n_permutations
  }

  list(
    variable_names = predictors,
    importance = as.numeric(importance_scores),
    baseline_score = baseline_r2,
    n_permutations = n_permutations,
    seed = random_seed
  )
}

# =============================================================================
# Main
# =============================================================================

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Set defaults
default_csv    <- "../../../datasets/csv/mtcars.csv"
default_output <- "../../../results/r"

# Parse arguments
csv_path_raw   <- ifelse(length(args) >= 1, args[1], default_csv)
output_dir_raw <- ifelse(length(args) >= 2, args[2], default_output)

# Resolve paths
csv_path   <- resolve_path(csv_path_raw)
output_dir <- resolve_path(output_dir_raw)

# Validate CSV path
if (!file.exists(csv_path)) {
  stop(paste("CSV file not found:", csv_path))
}

# Extract dataset name from filename
dataset_name <- tools::file_path_sans_ext(basename(csv_path))

cat(sprintf("Running feature importance test on dataset: %s\n", dataset_name))

# Load data
data <- read.csv(csv_path)
data <- convert_categorical_to_numeric(data, dataset_name)

# Use first column as response, rest as predictors
response <- names(data)[1]  # Get FIRST column (index 0 in 0-based, but R is 1-based)
predictors <- setdiff(names(data), response)

cat("Predictors:", paste(predictors, collapse = ", "), "\n")
cat("Response:", response, "\n")
cat("Observations:", nrow(data), "\n")

# Skip if too few predictors
if (length(predictors) < 1) {
  cat("Skipping: insufficient predictors\n")
  quit(status = 1)
}

n_features <- length(predictors)
has_vif <- n_features >= 2

# Build result object
result <- list(
  test = "feature_importance",
  method = "lm",
  dataset = dataset_name,
  n = nrow(data),
  k = n_features,
  variable_names = predictors,
  response = response
)

# 1. Standardized Coefficients
cat("\n[1/4] Standardized Coefficients...\n")
tryCatch({
  std_coefs <- validate_standardized_coefficients(data, predictors, response)
  result$standardized_coefficients <- std_coefs
  cat("   Variables:", paste(std_coefs$variable_names, collapse = ", "), "\n")
  cat("   Standardized Coefs:", paste(round(std_coefs$standardized_coefficients, 4), collapse = ", "), "\n")
}, error = function(e) {
  cat("   Error:", conditionMessage(e), "\n")
})

# 2. VIF Ranking (skip for single predictor)
if (has_vif) {
  cat("\n[2/4] VIF Ranking...\n")
  tryCatch({
    vif_result <- validate_vif_ranking(data, predictors, response)
    result$vif_ranking <- vif_result
    for (i in 1:min(5, nrow(vif_result$ranking))) {
      v <- vif_result$ranking[i, ]
      cat("   ", v$variable, ": VIF=", round(v$vif, 2), "(", v$interpretation, ")\n", sep="")
    }
  }, error = function(e) {
    cat("   Error:", conditionMessage(e), "\n")
  })
} else {
  cat("\n[2/4] VIF Ranking... Skipped (only 1 predictor)\n")
}

# 3. Linear SHAP
cat("\n[3/4] Linear SHAP...\n")
tryCatch({
  shap_result <- validate_linear_shap(data, predictors, response)
  result$shap <- shap_result
  cat("   Mean |SHAP|:", paste(round(shap_result$mean_abs_shap, 4), collapse = ", "), "\n")
}, error = function(e) {
  cat("   Error:", conditionMessage(e), "\n")
})

# 4. Permutation Importance
cat("\n[4/4] Permutation Importance...\n")
tryCatch({
  perm_result <- validate_permutation_importance(data, predictors, response, n_permutations = 10)
  result$permutation_importance <- perm_result
  cat("   Importance scores:", paste(round(perm_result$importance, 4), collapse = ", "), "\n")
}, error = function(e) {
  cat("   Error:", conditionMessage(e), "\n")
})

# Ensure output directory exists
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# Write output
output_file <- file.path(output_dir, paste0(dataset_name, "_feature_importance.json"))
write_json(result, output_file, pretty = TRUE, auto_unbox = FALSE, digits = 22)

cat(sprintf("\nWrote: %s\n", normalizePath(output_file)))
